// @tonaljs/rhythm-pattern/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// index.ts
var rhythm_patterm_exports = {};
__export(rhythm_patterm_exports, {
  binary: () => binary,
  euclid: () => euclid,
  hex: () => hex,
  onsets: () => onsets,
  probability: () => probability,
  random: () => random,
  rotate: () => rotate
});
module.exports = __toCommonJS(rhythm_patterm_exports);
function binary(...numbers) {
  return numbers.reduce((pattern, number) => {
    number.toString(2).split("").forEach((digit) => {
      pattern.push(parseInt(digit));
    });
    return pattern;
  }, []);
}
function hex(hexNumber) {
  const pattern = [];
  for (let i = 0; i < hexNumber.length; i++) {
    const digit = parseInt("0x" + hexNumber[i]);
    const binary2 = isNaN(digit) ? "0000" : digit.toString(2).padStart(4, "0");
    binary2.split("").forEach((digit2) => {
      pattern.push(digit2 === "1" ? 1 : 0);
    });
  }
  return pattern;
}
function onsets(...numbers) {
  return numbers.reduce((pattern, number) => {
    pattern.push(1);
    for (let i = 0; i < number; i++) {
      pattern.push(0);
    }
    return pattern;
  }, []);
}
function random(length, probability2 = 0.5, rnd = Math.random) {
  const pattern = [];
  for (let i = 0; i < length; i++) {
    pattern.push(rnd() >= probability2 ? 1 : 0);
  }
  return pattern;
}
function probability(probabilities, rnd = Math.random) {
  return probabilities.map((probability2) => rnd() <= probability2 ? 1 : 0);
}
function rotate(pattern, rotations) {
  const len = pattern.length;
  const rotated = [];
  for (let i = 0; i < len; i++) {
    const pos = ((i - rotations) % len + len) % len;
    rotated[i] = pattern[pos];
  }
  return rotated;
}
function euclid(steps, beats) {
  const pattern = [];
  let d = -1;
  for (let i = 0; i < steps; i++) {
    const v = Math.floor(i * (beats / steps));
    pattern[i] = v !== d ? 1 : 0;
    d = v;
  }
  return pattern;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  binary,
  euclid,
  hex,
  onsets,
  probability,
  random,
  rotate
});
